processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java

```diff
@@ -327,35 +345,46 @@ public static int getCardinalityForArrayAggregation(
   }
 
   /**
-   * Checks whether all "dimensions" are either single-valued, or if the input column or output dimension spec has
-   * specified a type that {@link ColumnType#isArray()}. Both cases indicate we don't want to explode the under-lying
-   * multi value column. Since selectors on non-existent columns will show up as full of nulls, they are effectively
-   * single valued, however capabilites on columns can also be null, for example during broker merge with an 'inline'
-   * datasource subquery, so we only return true from this method when capabilities are fully known.
+   * Returns all dimension names that are or could be multi valued, or if the input column specified a type that
+   * {@link ColumnType#isArray()}. Both cases indicate we want to explode the under-lying multi value column. Since
+   * selectors on non-existent columns will show up as full of nulls, they are effectively single valued, however
+   * capabilites on columns can also be null, for example during broker merge with an 'inline' datasource subquery. We
+   * mark columns with null capabilites as candidates for explosion.
    */
-  public static boolean hasNoExplodingDimensions(
+  public static List<String> findAllProbableExplodingDimensions(
```

Comment:
Hmm, looking through this I just realized that this function can't tell if dimensions are definitely multi-valued or not. It can only tell if they _might_ be multi-valued. (But they also might not be!)

So I think we should move the check; instead of checking here, we should check row by row as we actually run the query. If a multi-value row is ever encountered, then at that point we should throw the error.

Reply:
Thanks for the catch. Updated the check while doing row by row iteration.